File: cf_sdk/src/main/java/com/example/cf_sdk/logic/changebanksdk/source/AuthenticationDatasource.java
package com.example.cf_sdk.logic.changebanksdk.source;

import com.example.cf_sdk.logic.changebanksdk.model.FingerprintResponse;
import com.example.cf_sdk.logic.changebanksdk.model.Session;
import com.example.cf_sdk.logic.changebanksdk.model.member.SecurityProfile;
import com.example.cf_sdk.logic.changebanksdk.parameter.Parameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.CheckPasscodeEnabledParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.GetSecurityProfileParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.LoginParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.ResendTwoFactorParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.SecurityProfileParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.TwoFactorParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.UpdateCredentialsParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.UpdatePasswordParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.UsernameParameters;
import com.example.cf_sdk.logic.changebanksdk.response.ChangebankResponse;
import com.google.common.base.Optional;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 *
 * DataSource to provide authentication endpoints.
 */

public interface AuthenticationDatasource {
    Single<Session> login(LoginParameters loginParameters);

    Single<ChangebankResponse> updateCredentials(UpdateCredentialsParameters updateCredentialsParameters);

    Single<Session> getSession();

    Single<LoginParameters.LoginMethodType> getLoginMethodType();

    Single<Optional<byte[]>> getPinSalt();

    Single<ChangebankResponse> checkUsernameExists(UsernameParameters usernameParameters);

    Single<Optional<String>> getSavedUsername();

    Single<Optional<String>> getCachedUsername();

    Single<ChangebankResponse> logout(Parameters parameters);

    Single<ChangebankResponse> changePassword(UpdatePasswordParameters updatePasswordParameters);

    Single<Boolean> checkPasscodeEnabled(CheckPasscodeEnabledParameters parameters);

    void saveSession(Session session);

    void savePinSalt(byte[] pinSalt, String username);

    void saveUsername(String username);

    void clearAllAuthenticationDatasourceCache();

    void clearPinAndFingerprint();

    void updateLocalStorage(LoginParameters loginParameters);

    Observable<FingerprintResponse> readFingerprint(String username);

    Single<FingerprintResponse> readFingerprintInstant(String username);

    Completable writeFingerprint(String username, String passwordHash);

    Single<Boolean> canStoreFingerprintSecurely();

    Completable clearAllFingerprint();

    Single<Boolean> isUserFingerprintSaved(String username);

    Completable clearSpecificFingerprint(String username);

    Single<Session> resendLoginMFA(ResendTwoFactorParameters resendTwoFactorParameters);

    Single<Session> verifyLoginMFA(TwoFactorParameters twoFactorParameters);

    Single<ChangebankResponse> updateSecurityProfile(SecurityProfileParameters securityProfileParameters);

    Single<SecurityProfile> getSecurityProfile(GetSecurityProfileParameters getSecurityProfileParameters);

    Completable clearPasscode();
}

File: cf_sdk/src/main/java/com/example/cf_sdk/changebankapi/source/remote/AuthenticationDatasource.kt
package com.example.cf_sdk.changebankapi.source.remote


import com.example.cf_sdk.changebankapi.model.FingerprintResponse
import com.example.cf_sdk.changebankapi.model.Session
import com.example.cf_sdk.changebankapi.model.member.SecurityProfile
import com.example.cf_sdk.changebankapi.parameter.Parameters
import com.example.cf_sdk.changebankapi.parameter.authentication.GetSecurityProfileParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.LoginMethodType
import com.example.cf_sdk.changebankapi.parameter.authentication.LoginParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.ResendTwoFactorParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.SecurityProfileParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.TwoFactorParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.UsernameParameters
import com.example.cf_sdk.changebankapi.response.ChangebankResponse

import com.google.common.base.Optional
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single

/**
 * Created by victorojeda on 10/19/16.
 *
 *
 * DataSource to provide authentication endpoints.
 */
interface AuthenticationDatasource {
    val session: Single<Session?>?
    val loginMethodType: Single<LoginMethodType?>?
    val pinSalt: Single<Optional<ByteArray?>?>?
    fun checkUsernameExists(usernameParameters: UsernameParameters?): Single<ChangebankResponse?>?
    val savedUsername: Single<Optional<String?>?>?
    val cachedUsername: Single<Optional<String?>?>?
    fun logout(parameters: Parameters?): Single<ChangebankResponse?>?
    fun saveSession(session: Session?)
    fun savePinSalt(pinSalt: ByteArray?, username: String?)
    fun saveUsername(username: String?)
    fun clearAllAuthenticationDatasourceCache()
    fun clearPinAndFingerprint()
    fun updateLocalStorage(loginParameters: LoginParameters?)
    fun readFingerprint(username: String?): Observable<FingerprintResponse?>?
    fun readFingerprintInstant(username: String?): Single<FingerprintResponse?>?
    fun writeFingerprint(username: String?, passwordHash: String?): Completable?
    fun canStoreFingerprintSecurely(): Single<Boolean?>?
    fun clearAllFingerprint(): Completable?
    fun isUserFingerprintSaved(username: String?): Single<Boolean?>?
    fun clearSpecificFingerprint(username: String?): Completable?
    fun resendLoginMFA(resendTwoFactorParameters: ResendTwoFactorParameters?): Single<Session?>?
    fun verifyLoginMFA(twoFactorParameters: TwoFactorParameters?): Single<Session?>?
    fun updateSecurityProfile(securityProfileParameters: SecurityProfileParameters?): Single<ChangebankResponse?>?
    fun getSecurityProfile(getSecurityProfileParameters: GetSecurityProfileParameters?): Single<SecurityProfile?>?
    fun clearPasscode(): Completable?
}